devito/types/basic.py

```diff
@@ -416,6 +425,114 @@ def is_const(self):
+    """
+    Base class for vector and tensor valued functions. It inherits from and
+    mimicks the behavior of a sympy.ImmutableDenseMatrix
+
+    The sub-hierachy is as follows
+
+                         AbstractTensor
+                                |
+                      AbstractCachedTensor
+                                |
+                          TensorFunction
+                                |
+                 ---------------------------------
+                 |                               |
+          VectorFunction                 TensorTimeFunction
+                        \-------\                |
+                                 \------- VectorimeFunction
+
+    There are four relevant AbstractFunction sub-types: ::
+
+        * TensorFunction: A space-varying vector valued function
+        * VectorFunction: A space-varying tensor valued function
+        * TensorTimeFunction: A time-space-varying vector valued function
+        * VectorTImeFunction: A time-space-varying tensor valued function
+    """
+
+    is_AbstractTensor = True
+
+
+class AbstractCachedTensor(AbstractTensor, Cached, Evaluable):
+    """
+    Base class for tensor-valued tensor symbols, cached by both Devito and Sympy.
+
+    For more information, refer to ``AbstractTensor.__doc__``.
+    """
+    is_MatrixLike = True
+    is_Matrix = True
+    is_TensorValued = True
+    is_VectorValued = False
+
+    def __new__(cls, *args, **kwargs):
+        options = kwargs.get('options', {})
+        if cls in _SymbolCache:
+            newobj = sympy.Matrix.__new__(cls, *args, **options)
+            newobj._cached_init()
+        else:
+            name = kwargs.get('name')
+            # Number of dimensions
+            indices, _ = cls.__indices_setup__(**kwargs)
+
+            # Create the new Function object and invoke __init__
+            comps = cls.__setup_subfunc__(*args, **kwargs)
+            newcls = cls._symbol_type(name)
+            newobj = sympy.ImmutableDenseMatrix.__new__(newcls, comps)
+            # Initialization. The following attributes must be available
+            newobj._indices = indices
+            newobj._name = name
+            newobj._dtype = cls.__dtype_setup__(**kwargs)
+            newobj.__init__(*args, **kwargs)
+            # Store new instance in symbol cache
+            newobj._cache_put(newobj)
+
+        return newobj
+
+    __hash__ = Cached.__hash__
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            pass
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        """Extract the object data type from ``kwargs``."""
+        return None
+
+    @classmethod
+    def __setup_subfunc__(cls, *args, **kwargs):
+        """
+        Setup each component as a Devito type
+        """
+        return []
+
+    @classmethod
+    def __indices_setup__(cls, *args, **kwargs):
+        """
+        Setup each component as a Devito type
```

Comment:
full stop at the end. 

We also need to document why the return value is twofold.

Also, it's time to move this `__indices_setup__`, along with `__shape_setup__` and its siblings, to a common superclass (our own `Basic` ?). Otherwise, it gets duplicated across `AbstractTensor`, `AbstractSymbol`, and `AbstractFunction` !